adapters/beachfront/beachfront.go

```diff
@@ -111,427 +96,588 @@ type BeachfrontResponseSlot struct {
+				H: request.Imp[i].Banner.Format[j].H,
+				W: request.Imp[i].Banner.Format[j].W,
+			})
+		}
+
+		bfr.Slots = append(bfr.Slots, slot)
 	}
 
-	// Just take the last one... I guess?
-	if req.Imp[bannerImpsIndex].Secure != nil {
-		beachfrontReq.Secure = *req.Imp[bannerImpsIndex].Secure
+	if len(bfr.Slots) == 0 {
+		return bfr, errs
 	}
 
-	if req.User != nil {
-		beachfrontReq.User.ID = req.User.ID
-		beachfrontReq.User.BuyerUID = req.User.BuyerUID
+	if request.Device != nil {
+		bfr.IP = getIP(request.Device.IP)
+		bfr.DeviceModel = request.Device.Model
+		bfr.DeviceOs = request.Device.OS
+		if request.Device.DNT != nil {
+			bfr.Dnt = *request.Device.DNT
+		}
+		if request.Device.UA != "" {
+			bfr.UA = request.Device.UA
+		}
 	}
 
-	if req.App != nil {
-		beachfrontReq.Domain = req.App.Domain
-		beachfrontReq.Page = req.App.ID
-	} else {
-		protoUrl := strings.Split(req.Site.Page, "//")
-		var domainPage string
-		// Resolves a panic for any Site.Page that does not include the protocol
-		if len(protoUrl) > 1 {
-			domainPage = protoUrl[1]
+	var t = fallBackDeviceType(request)
+
+	if t == openrtb.DeviceTypeMobileTablet {
+		bfr.Page = request.App.Bundle
+		if request.App.Domain == "" {
+			bfr.Domain = getDomain(request.App.Domain)
+		} else {
+			bfr.Domain = request.App.Domain
+		}
+
+		bfr.IsMobile = 1
+	} else if t == openrtb.DeviceTypePersonalComputer {
+		bfr.Page = request.Site.Page
+		if request.Site.Domain == "" {
+			bfr.Domain = getDomain(request.Site.Page)
 		} else {
-			domainPage = protoUrl[0]
+			bfr.Domain = request.Site.Domain
+		}
+
+		bfr.IsMobile = 0
+	}
+
+	bfr.Secure = isSecure(bfr.Page)
+
+	if request.User != nil && request.User.ID != "" {
+		if bfr.User.ID == "" {
+			bfr.User.ID = request.User.ID
 		}
-		beachfrontReq.Domain = strings.Split(domainPage, "/")[0]
-		beachfrontReq.Page = req.Site.Page
 	}
 
-	beachfrontReq.RequestID = req.ID
+	if request.User != nil && request.User.BuyerUID != "" {
+		if bfr.User.BuyerUID == "" {
+			bfr.User.BuyerUID = request.User.BuyerUID
+		}
+	}
+
+	bfr.RequestID = request.ID
+	bfr.AdapterName = beachfrontAdapterName
+	bfr.AdapterVersion = beachfrontAdapterVersion
+
+	if request.Imp[0].Secure != nil {
+		bfr.Secure = *request.Imp[0].Secure
+	}
+
+	return bfr, errs
+}
+
+func fallBackDeviceType(request *openrtb.BidRequest) openrtb.DeviceType {
+	if request.Site != nil {
+		return openrtb.DeviceTypePersonalComputer
+	}
 
-	return beachfrontReq, errs, imps
+	return openrtb.DeviceTypeMobileTablet
 }
 
-func getVideoRequest(req *openrtb.BidRequest) (BeachfrontVideoRequest, []error, int) {
-	var videoImpsIndex = 0
-	var beachfrontReq = NewBeachfrontVideoRequest()
-	var errs = make([]error, 0, len(req.Imp))
-	var imps = 0
+func getVideoRequests(request *openrtb.BidRequest) ([]beachfrontVideoRequest, []error) {
+	var bfReqs = make([]beachfrontVideoRequest, len(request.Imp))
+	var errs = make([]error, 0, len(request.Imp))
+	var bad = make([]int, 0)
```

Comment:
Based on usage, consider renaming to something along the lines of `failedRequesIndices`.